### PR TITLE
FIX: segfault on truncated png

### LINKED ISSUE #9256: reading truncated png can segfault python
### Bug report

**Bug summary**

Reading a truncated png file can segfault the process.

**Code for reproduction**

```python
from pylab import *
rcdefaults()
savefig("test.png")
buf = open("test.png", "rb").read()
open("test.png", "wb").write(buf[:20])  # truncate to first 20 bytes
imread("test.png")
```

**Actual outcome**

```
Fatal Python error: Segmentation fault

Current thread 0x00007f3638e8a540 (most recent call first):
  File "/usr/lib/python3.6/site-packages/matplotlib/image.py", line 1298 in imread
  File "/usr/lib/python3.6/site-packages/matplotlib/pyplot.py", line 2314 in imread
  File "test.py", line 6 in <module>
[1]    12789 segmentation fault (core dumped)  python test.py
```

**Expected outcome**

Some exception is raised.

(To be honest I'm not fully convinced there's a much of a benefit in shipping our own png wrappers rather than, say, relying on PIL throughout.)

**Matplotlib version**

  * Operating System: Arch Linux
  * Matplotlib Version: 2.0.2 and master
  * Python Version: 3.6
  * Jupyter Version (if applicable): NA
  * Other Libraries: NA

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary
 - be more paranoid in _read_png_data about failed reads and check for
   exceptions after both calls to it
 - in read_png_data kick the png error handling
 - only override the exception in the body of `setjmp` handler if
   there is not already one set

closes #9256

## PR Checklist

- [x] Has Pytest style unit tests

OP has a test case, just need to add it.

### CODE REVIEW

File: src/_png.cpp [hunk 1]
```diff
@@ -408,13 +408,21 @@ static void _read_png_data(PyObject *py_file_obj, png_bytep data, png_size_t len
...
     if (read_method) {
         result = PyObject_CallFunction(read_method, (char *)"i", length);
-        if (PyBytes_AsStringAndSize(result, &buffer, &bufflen) == 0) {
-            if (bufflen == (Py_ssize_t)length) {
-                memcpy(data, buffer, length);
-            } else {
-                PyErr_SetString(PyExc_IOError, "read past end of file");
-            }
-        }
+	if (result) {
```
Comment (QuLogic) [inline src/_png.cpp]: All the code uses some mixed tab/space indent. It should be 4-space indent even in the C code, I believe.

File: src/_png.cpp [hunk 2]
```diff
@@ -408,13 +408,21 @@ static void _read_png_data(PyObject *py_file_obj, png_bytep data, png_size_t len
...
+	    if (PyBytes_AsStringAndSize(result, &buffer, &bufflen) == 0) {
+		if (bufflen == (Py_ssize_t)length) {
+		    memcpy(data, buffer, length);
+		} else {
+		    PyErr_SetString(PyExc_IOError, "read past end of file");
+		}
+	    } else {
+		PyErr_SetString(PyExc_IOError, "Failed to copy buffer");
+	    }
+	} else 	{
```
Comment (QuLogic) [inline src/_png.cpp]: Weird tab in the middle there.

File: src/_png.cpp [hunk 3]
```diff
@@ -424,6 +432,11 @@ static void read_png_data(png_structp png_ptr, png_bytep data, png_size_t length
 {
     PyObject *py_file_obj = (PyObject *)png_get_io_ptr(png_ptr);
     _read_png_data(py_file_obj, data, length);
+    if (PyErr_Occurred())
+    {
```
Comment (QuLogic) [inline src/_png.cpp]: Brace goes on `if` line.

File: src/_png.cpp [hunk 4]
```diff
@@ -481,6 +494,9 @@ static PyObject *_read_png(PyObject *filein, bool float_result)
         }
         Py_XDECREF(read_method);
         _read_png_data(py_file, header, 8);
+	if (PyErr_Occurred()){
```
Comment (QuLogic) [inline src/_png.cpp]: Space before brace.

File: src/_png.cpp [hunk 5]
```diff
@@ -503,7 +519,10 @@ static PyObject *_read_png(PyObject *filein, bool float_result)
     }
 
     if (setjmp(png_jmpbuf(png_ptr))) {
-        PyErr_SetString(PyExc_RuntimeError, "Error setting jump");
+	if (!PyErr_Occurred())
+	{
```
Comment (QuLogic) [inline src/_png.cpp]: Brace goes on `if` line.

File: src/_png.cpp [hunk 6]
```diff
@@ -424,6 +432,11 @@ static void read_png_data(png_structp png_ptr, png_bytep data, png_size_t length
 {
     PyObject *py_file_obj = (PyObject *)png_get_io_ptr(png_ptr);
     _read_png_data(py_file_obj, data, length);
+    if (PyErr_Occurred())
```
Comment (anntzer) [inline src/_png.cpp]: -Wmisleading-indentation :-) (here and below)

**Comment by tacaswell** [comment]
I do not do a lot of c-code and the auto-formatter in emacs was fighting back....

File: lib/matplotlib/tests/test_png.py [hunk 7]
```diff
@@ -46,3 +47,30 @@ def test_imread_png_uint16():
...
+    plt.savefig(fname)
+    with open(fname, 'rb') as fin:
+        buf = fin.read()
+    with open(fname_t, 'wb') as fout:
+        fout.write(buf[:20])
+
+    with pytest.raises(Exception):
+        plt.imread(fname_t)
+
+    shutil.rmtree(d)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_png.py]: This should be in `finally`, or maybe use the `tempdir` pytest fixture.

File: lib/matplotlib/tests/test_png.py [hunk 8]
```diff
@@ -46,3 +47,30 @@ def test_imread_png_uint16():
 
     assert (img.dtype == np.uint16)
     assert np.sum(img.flatten()) == 134184960
+
+
+def test_truncated_file():
```
Comment (anntzer) [inline lib/matplotlib/tests/test_png.py]: Do we really need to test both filenames and streams?  (or rather, there doesn't seem to be a test of successfully reading from a stream (as opposed to a filename), which seems something more valuable to check...)

**Comment by tacaswell** [comment]
@antongulikov why?

**Comment by anntzer** [comment]
If anything, error messages should start with a lower case letter and not have final dots for consistency with CPython itself IMO... (e.g., `list index out of range`, `type object 'Foo' has no attribute 'bar`)

File: lib/matplotlib/tests/test_png.py (hunk 8 above)
Comment (tacaswell) [reply lib/matplotlib/tests/test_png.py]: Yes, because they go through different code paths.

Comment (anntzer) [reply lib/matplotlib/tests/test_png.py]: but not in the erroring part right?  (at that point every possible input has been converted to an object with a read method)

Comment (tacaswell) [reply lib/matplotlib/tests/test_png.py]: :sheep: Yup, I was confused be the `_read_png_data` in the first if / else block and thought that was reading the whole file, but is just pulling off the first 8 bytes.

**Comment by antongulikov** [comment]
@tacaswell I wanted to say that the style for error messages should be the same for different exceptions.

**Comment by tacaswell** [comment]
@antongulikov That makes sense, I think we have made the all consistent now.

**Pull request merged**